file_store.go

```diff
@@ -0,0 +1,208 @@
+		if err != nil {
+			return indexes, fmt.Errorf("Unexpected filename, does not confirm to entry.%%d: %q. Parse error: %v", name, err)
+		}
+
+		indexes = append(indexes, uint64(index))
+	}
+
+	sort.Sort(uint64Slice(indexes))
+
+	return indexes, nil
+}
+
+// FirstIndex implements the LogStore interface.
+func (s *FileStore) FirstIndex() (uint64, error) {
+	s.l.RLock()
+	defer s.l.RUnlock()
+	return s.lowIndex, nil
+}
+
+// LastIndex implements the LogStore interface.
+func (s *FileStore) LastIndex() (uint64, error) {
+	s.l.RLock()
+	defer s.l.RUnlock()
+	return s.highIndex, nil
+}
+
+// GetLog implements the LogStore interface.
+func (s *FileStore) GetLog(index uint64, rlog *Log) error {
+	s.l.RLock()
+	defer s.l.RUnlock()
+	f, err := os.Open(filepath.Join(s.dir, "logs/entry."+strconv.FormatUint(index, 10)))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return ErrLogNotFound
+		}
+		return err
+	}
+	defer f.Close()
+
+	var elog Log
+	if err := gob.NewDecoder(f).Decode(&elog); err != nil {
+		return err
+	}
+	*rlog = elog
+	return nil
+}
+
+// StoreLog implements the LogStore interface.
+func (s *FileStore) StoreLog(log *Log) error {
+	return s.StoreLogs([]*Log{log})
+}
+
+// StoreLogs implements the LogStore interface.
+func (s *FileStore) StoreLogs(logs []*Log) error {
+	s.l.Lock()
+	defer s.l.Unlock()
+
+	for _, entry := range logs {
+		f, err := os.Create(filepath.Join(s.dir, "logs/entry."+strconv.FormatUint(entry.Index, 10)))
+		if err != nil {
+			return err
+		}
+		if err := gob.NewEncoder(f).Encode(entry); err != nil {
+			f.Close()
+			return err
+		}
+		if entry.Index < s.lowIndex || s.lowIndex == 0 {
+			s.lowIndex = entry.Index
+		}
+		if entry.Index > s.highIndex {
+			s.highIndex = entry.Index
+		}
+		f.Close()
+	}
+
+	return nil
+}
+
+// DeleteRange implements the LogStore interface.
+func (s *FileStore) DeleteRange(min, max uint64) error {
+	s.l.Lock()
+	defer s.l.Unlock()
+	for i := min; i <= max; i++ {
+		if err := os.Remove(filepath.Join(s.dir, "logs/entry."+strconv.FormatUint(i, 10))); err != nil {
+			return err
+		}
+	}
+	s.lowIndex = max + 1
+	return nil
+}
+
+func (s *FileStore) DeleteAll() error {
+	s.l.Lock()
+	defer s.l.Unlock()
+	if err := os.RemoveAll(filepath.Join(s.dir, "logs")); err != nil {
+		return err
+	}
+	if err := os.MkdirAll(filepath.Join(s.dir, "logs"), 0700); err != nil {
+		return err
+	}
+	s.lowIndex = 0
+	s.highIndex = 0
+	return nil
+}
+
+// Set implements the StableStore interface.
+func (s *FileStore) Set(key []byte, val []byte) error {
+	s.l.Lock()
+	defer s.l.Unlock()
+	return ioutil.WriteFile(filepath.Join(s.dir, "stable", "key."+string(key)), val, 0600)
+}
+
+// Get implements the StableStore interface.
+func (s *FileStore) Get(key []byte) ([]byte, error) {
+	s.l.RLock()
+	defer s.l.RUnlock()
+	b, err := ioutil.ReadFile(filepath.Join(s.dir, "stable", "key."+string(key)))
+	if err != nil && os.IsNotExist(err) {
+		return []byte{}, fmt.Errorf("not found")
```

Comment:
Sent https://github.com/hashicorp/raft/pull/33 to clarify that in the documentation.